File: python/n447_Boomerangs.py
# https://leetcode-cn.com/problems/number-of-boomerangs/
class Solution:
    def numberOfBoomerangs(self, points) -> int:
        dist_table = {}
        res = 0
        for i in range(len(points)):
            for j in range(len(points)):
                dist_two = self.dist(points[i], points[j])
                if  dist_two in dist_table.keys():
                    dist_table[dist_two] += 1
                else:
                    dist_table[dist_two] = 1
            for i in dist_table.values():
                res += i * (i-1)
            dist_table.clear()
        return res

    def dist(self, point_a, point_b):
        return (point_a[0] - point_b[0]) ** 2 + (point_a[1] - point_b[1]) ** 2
        
        
if __name__ == "__main__":
    pass

File: cpp/n447_Boomerangs.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int d2(int x1, int x2, int y1, int y2) {
        int d1 = x1 - x2;
        int d2 = y1 - y2;
        return d1 * d1 + d2 * d2;
    }
    int numberOfBoomerangs(vector<vector<int>>& points) {
        int N = points.size();
        vector<unordered_map<int, int> > dists(N);
        int res = 0;
        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < i; ++j) {
                int d = d2(points[i][0], points[j][0], points[i][1], points[j][1]);
                res += 2 * (dists[i][d] + dists[j][d]);
                ++dists[i][d];
                ++dists[j][d];
            }
        }
        return res;
    }
};

int main(){
    return 0;
}